### PR TITLE
[CALCITE-3644] Add ProjectSetOpTransposeRule to normalize materialized view

### PR DESCRIPTION
**Pull request by xy2953396112**
JIRA:https://issues.apache.org/jira/browse/CALCITE-3644

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/RelOptMaterializations.java [hunk 1]
```diff
@@ -201,6 +202,7 @@
             .addRuleInstance(ProjectMergeRule.INSTANCE)
             .addRuleInstance(ProjectRemoveRule.INSTANCE)
             .addRuleInstance(ProjectJoinTransposeRule.INSTANCE)
+            .addRuleInstance(ProjectSetOpTransposeRule.INSTANCE)
             .addRuleInstance(FilterToCalcRule.INSTANCE)
             .addRuleInstance(ProjectToCalcRule.INSTANCE)
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/plan/RelOptMaterializations.java]: LGTM

**Comment by xy2953396112** [comment]
@jinxing64 Help review this pr when you have time.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 2]
```diff
@@ -1621,8 +1621,31 @@ public UnifyResult apply(UnifyRuleCall call) {
             }
           }
         }
+
+        // try to adjust RelDataType
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java]: What's this change for ?
Do we have a test ?

File: core/src/main/java/org/apache/calcite/plan/RelOptMaterializations.java [hunk 3]
```diff
@@ -201,6 +202,7 @@
             .addRuleInstance(ProjectMergeRule.INSTANCE)
             .addRuleInstance(ProjectRemoveRule.INSTANCE)
             .addRuleInstance(ProjectJoinTransposeRule.INSTANCE)
+            .addRuleInstance(ProjectSetOpTransposeRule.INSTANCE)
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/plan/RelOptMaterializations.java]: +1 for this change.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 4]
```diff
@@ -1621,8 +1621,31 @@ public UnifyResult apply(UnifyRuleCall call) {
...
+        List<RexNode> projectExprs = new ArrayList<>();
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java]: Use `final` ?

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java (hunk 2 above)
Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java]: Can we doc the change ?

Comment (xy2953396112) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java]: > What's this change for ?
> Do we have a test ?

@jinxing64 Thanks for you review, this change is used for adjusting RelDataType.
Such as  `java.lang.string` and  `VARCHAR`. Left comment in jira yesterday, Maybe we can discuss it in [jira](https://issues.apache.org/jira/browse/CALCITE-3644)

**Comment by xy2953396112** [comment]
@jinxing64 commented in [JIRA](https://issues.apache.org/jira/browse/CALCITE-3644). May not be appropriate in the current implementation. Would you take a look, please ?

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java (hunk 4 above)
Comment (xy2953396112) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java]: Thanks for you review.  create `createProjects` to adjust RelDataType.

**Comment by xy2953396112** [comment]
> @jinxing64 I have updated codes, please help review this pr when you have time, thanks.

**Comment by xy2953396112** [comment]
> +1

Thanks

File: core/src/main/java/org/apache/calcite/rel/mutable/MutableRels.java [hunk 5]
```diff
@@ -173,6 +173,23 @@ public RexNode get(int index) {
         .collect(Collectors.toList());
   }
 
+  /**
+   * Construct expression list of Project by the given fields of the input.
+   */
+  public static List<RexNode> createProjects(final MutableRel child,
+      final List<RexNode> projs) {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/mutable/MutableRels.java]: What does it do?
Is it really a public API?

What if I pass `+($0, 1)` as a project expression what would it do?

File: core/src/main/java/org/apache/calcite/rel/mutable/MutableRels.java [hunk 6]
```diff
@@ -173,6 +173,23 @@ public RexNode get(int index) {
...
+    List<RexNode> rexNodeList = new ArrayList<>();
+    for (int i = 0; i < projs.size(); i++) {
+      if (projs.get(i) instanceof RexInputRef) {
+        RexInputRef rexInputRef = (RexInputRef) projs.get(i);
+        rexNodeList.add(RexInputRef.of(rexInputRef.getIndex(), child.rowType));
+      } else {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/mutable/MutableRels.java]: What if `RexInputRef` is located inside a `RexCall`?

File: core/src/main/java/org/apache/calcite/rel/mutable/MutableRels.java [hunk 7]
```diff
@@ -173,6 +173,23 @@ public RexNode get(int index) {
         .collect(Collectors.toList());
   }
 
+  /**
+   * Construct expression list of Project by the given fields of the input.
+   */
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/rel/mutable/MutableRels.java]: Seems this method only converts the row type for `RexInputRef`, why this is taken into consideration specially ?
BTW I can pass the test you provided in MaterializationTest.java without this change.

**Pull request closed without merging**